app/javascript/controllers/select_with_auto_complete_controller.js

```diff
@@ -0,0 +1,499 @@
+    for (let i = 0; i < categories.length; i++) {
+      const category = categories[i];
+      this.#removeOptionEvents(category);
+    }
+  }
+
+  #attachOptionEvents(category, add = false) {
+    const categoryItems = category.querySelectorAll(':scope > [role="option"]');
+    for (let i = 0; i < categoryItems.length; i++) {
+      const categoryItem = categoryItems[i];
+      if (add) {
+        this.allOptions.push(categoryItem);
+      }
+      this.#addListboxOptionEventListeners(categoryItem);
+    }
+  }
+
+  #removeOptionEvents(category) {
+    const categoryItems = category.querySelectorAll(':scope > [role="option"]');
+    for (let i = 0; i < categoryItems.length; i++) {
+      const categoryItem = categoryItems[i];
+      this.#removeListboxOptionEventListeners(categoryItem);
+    }
+  }
+
+  #getLowercaseContent(node) {
+    return node.textContent.toLowerCase();
+  }
+
+  #isOptionInView(option) {
+    const bounding = option.getBoundingClientRect();
+    return (
+      bounding.top >= 0 &&
+      bounding.left >= 0 &&
+      bounding.bottom <=
+        (window.innerHeight || document.documentElement.clientHeight) &&
+      bounding.right <=
+        (window.innerWidth || document.documentElement.clientWidth)
+    );
+  }
+
+  #setActiveDescendant(option) {
+    if (option) {
+      this.comboboxTarget.setAttribute("aria-activedescendant", option.id);
+      if (!this.#isOptionInView(option)) {
+        option.scrollIntoView({ behavior: "smooth", block: "nearest" });
+      }
+    } else {
+      this.comboboxTarget.removeAttribute("aria-activedescendant");
+    }
+  }
+
+  #setValue(option) {
+    this.hiddenTarget.value = option ? option.getAttribute("data-value") : "";
+    this.filter = option ? option.textContent : "";
+    this.comboboxTarget.value = this.filter;
+    this.comboboxTarget.setSelectionRange(
+      this.filter.length,
+      this.filter.length,
+    );
+    this.#filterOptions();
+    this.#setOption(null);
+  }
+
+  #renderNoResults() {
+    const noResultsOption = document.createElement("div");
+    noResultsOption.setAttribute("role", "option");
+    noResultsOption.setAttribute("aria-disabled", "true");
+    noResultsOption.className =
+      "px-3 py-2 text-sm text-slate-500 dark:text-slate-300";
+    noResultsOption.textContent =
+      this.hasNoResultsTextValue && this.noResultsTextValue
+        ? this.noResultsTextValue
+        : "No results found";
+    this.listboxTarget.appendChild(noResultsOption);
+  }
+
+  // ComboboxAutocomplete events
+
+  #escapeRegExp(string) {
+    return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
+  }
+
+  #filterOptions() {
+    let option = null;
+    const currentOption = this.option;
+    const filter = this.filter.toLowerCase();
+
+    this.filteredOptions = [];
+    this.listboxTarget.innerHTML = "";
+
+    for (let i = 0; i < this.allOptions.length; i++) {
+      let flag = false;
+      const optionCategory = this.allOptions[i].cloneNode(true);
```

Comment:
This can be expensive, but I do understand why you are doing it.  The other option is to show/hide the existing options, but then we have to update the `mark`.